sympy/integrals/prde.py

```diff
@@ -359,35 +379,388 @@ def prde_no_cancel_b_small(b, Q, n, DE):
         c = eye(m)
         A = A.row_join(zeros(A.rows, m)).col_join(c.row_join(-c))
         return (H, A)
+
+    # else: b is in k, deg(qi) < deg(Dt)
+
+    t = DE.t
+    if DE.case != 'base':
+        with DecrementLevel(DE):
+            t0 = DE.t  # k = k0(t0)
+            ba, bd = frac_in(b, t0, field=True)
+            Q0 = [frac_in(qi.TC(), t0, field=True) for qi in Q]
+            f, B = param_rischDE(ba, bd, Q0, DE)
+
+            # f = [f1, ..., fr] in k^r and B is a matrix with
+            # m + r columns and entries in Const(k) = Const(k0)
+            # such that Dy0 + b*y0 = Sum(ci*qi, (i, 1, m)) has
+            # a solution y0 in k with c1, ..., cm in Const(k)
+            # if and only y0 = Sum(dj*fj, (j, 1, r)) where
+            # d1, ..., dr ar in Const(k) and
+            # B*Matrix([c1, ..., cm, d1, ..., dr]) == 0.
+
+        # Transform fractions (fa, fd) in f into constant
+        # polynomials fa/fd in k[t].
+        # (Is there a better way?)
+        f = [Poly(fa.as_expr()/fd.as_expr(), t, field=True)
+             for fa, fd in f]
     else:
-        # TODO: implement this (requires recursive param_rischDE() call)
-        raise NotImplementedError
+        # Base case. Dy == 0 for all y in k and b == 0.
+        # Dy + b*y = Sum(ci*qi) is solvable if and only if
+        # Sum(ci*qi) == 0 in which case the solutions are
+        # y = d1*f1 for f1 = 1 and any d1 in Const(k) = k.
+
+        f = [Poly(1, t, field=True)]  # r = 1
+        B = Matrix([[qi.TC() for qi in Q] + [S(0)]])
+        # The condition for solvability is
+        # B*Matrix([c1, ..., cm, d1]) == 0
+        # There are no constraints on d1.
+
+    # Coefficients of t^j (j > 0) in Sum(ci*qi) must be zero.
+    d = max([qi.degree(DE.t) for qi in Q])
+    if d > 0:
+        M = Matrix(d, m, lambda i, j: Q[j].nth(i + 1))
+        A, _ = constant_system(M, zeros(d, 1), DE)
+    else:
+        # No constraints on the hj.
+        A = Matrix(0, m, [])
+
+    # Solutions of the original equation are
+    #    y = Sum(dj*fj, (j, 1, r)) + Sum(ei*hi, (i, 1, m)),
+    # where  ei == ci  (i = 1, ..., m),  when
+    # A*Matrix([c1, ..., cm]) == 0 and
+    # B*Matrix([c1, ..., cm, d1, ..., dr]) == 0
+
+    # Build combined constraint matrix with m + r + m columns.
+
+    r = len(f)
+    I = eye(m)
+    A = A.row_join(zeros(A.rows, r + m))
+    B = B.row_join(zeros(B.rows, m))
+    C = I.row_join(zeros(m, r)).row_join(-I)
+
+    return f + H, A.col_join(B).col_join(C)
+
+
+def prde_cancel_primitive(b, Q, n, DE):
+    """
+    Pg, 237
+    """
+    m = len(Q)
+    f, A = [None]*(n + 1), [None]*(n + 1)
+
+    # Line from book: assuming we can solve such problems(prde's)
+    # over k. That is why we are using DecrementLevel.
+    with DecrementLevel(DE):
+        t0 = DE.t
+        ba, bd = frac_in(b, t0)
```

Comment:
I put this out of the loop, since it is just going to remain same. Only review this part and the first loop below it for now.